metadata/metadata.go

```diff
@@ -193,9 +193,7 @@ func FromIncomingContext(ctx context.Context) (MD, bool) {
 	}
 	out := make(MD, len(md))
 	for k, v := range md {
-		// We need to manually convert all keys to lower case, because MD is a
-		// map, and there's no guarantee that the MD attached to the context is
-		// created using our helper functions.
+		// Convert keys to lower case: MD can be modified after NewIncomingContext().
```

Comment:
I don't feel strongly here, but the intent of the original comment was to say that .. if only our methods are used to create/modify metadata, then everything is always lowercase.  But users can just do `metadata.NewOutgoingContext(ctx, metadata.MD{"mY-KeY": []string{"v"})` and we can't prevent that.

MD should not be modified after `NewIncomingContext`.  (Actually, users shouldn't need to call `NewIncomingContext` in the first place.)  It also shouldn't be modified after calling `NewOutgoingContext`.  I thought we documented this long ago, but I don't see it now...  If it's mutated in place then very bad things could happen.

Reply:
I've added a sentence to NewIncomingContext and NewOutgoingContext. I also thought I had seen that documented at some point.

I reverted this comment change. The reason this confused me is that the only way the `MD` can get in the context is via `NewIncomingContext` and `NewOutgoingContext`, so those functions could ensure the `MD` is in the correct form. The problem with that is since MD is a mutable map, it could be modified later.

I'm going to experiment with this a bit and I'll come back with an issue if I think there is something that is "easy" to change here that would help avoid copies.